### PR TITLE
3.1 passport hotfixes

### LINKED ISSUE #1136: TR1X bug: Game crashes when trying to deselect Load Game
Steps to reproduce:

- Press F6 to open the Load Game page
- Press Esc

20240115_141048_Caves

### LINKED ISSUE #1137: TR1X bug: Load Game text briefly appear when opening the passport if no saves are present
This bug happens on the Main Menu and inventory when opening the passport, if no saves are present.

I've confirmed this bug to affect version 3.1 but not 3.0.5.

### LINKED ISSUE #1140: Can't save in an empty slot using quick save if the load game menu was opened before

### LINKED ISSUE #1141: Show empty passport menu with quick load instead of opening the save game menu when there are no saves

### PR DESCRIPTION
**Pull request by walkawayy**
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description
Includes multiple passport fixes, a small refactor, and a change to quick load when there are no saves. Changes are best viewed on a commit by commit basis with their corresponding commit message and changelog entry. @aredfan tested the changes and had positive feedback. This will require a 3.1.1 hotfix since this fixes a game crash and a quick saving bug.

### CODE REVIEW

**Comment by rr-** [comment]
Thanks for investigating and quick fixes. In 3.0.5 and TombATI, if there were no saves present, pressing F6 would open a new game page. On these grounds I don't think we should disable F6 altogether, especially that the new approach makes it no longer possible to press F6 also in Gym. Below I propose a patch that restores the game modes dialog, along with opening automatically its requestor rather than requiring the player to press menu confirm button:

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 1d51eecd..bcf04676 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,5 @@
 ## [Unreleased](https://github.com/LostArtefacts/TR1X/compare/stable...develop) - ××××-××-××
-- disabled quick load if there are no saves present instead of opening the save game menu (#1141)
+- fixed quick load not showing New Game dialog if there are no saves present (#1141)
 - fixed a game crash when the quick load passport is deselected (#1136, regression from 3.1)
 - fixed not being able to save in an empty slot using quick save if the load game menu was opened before (#1140, regression from 3.1)
 - fixed the passport briefly flashing inaccessible page text (#1137, regression from 3.1)
diff --git a/src/game/game/game.c b/src/game/game/game.c
index d4b17900..0fe6309d 100644
--- a/src/game/game/game.c
+++ b/src/game/game/game.c
@@ -87,9 +87,7 @@ static int32_t Game_Control(int32_t nframes, GAMEFLOW_LEVEL_TYPE level_type)
                 g_OverlayFlag = 0;
             } else if (g_OverlayFlag > 0) {
                 if (g_Input.load) {
-                    if (g_SavedGamesCount > 0) {
-                        g_OverlayFlag = -1;
-                    }
+                    g_OverlayFlag = -1;
                 } else if (g_Input.save) {
                     g_OverlayFlag = -2;
                 } else {
diff --git a/src/game/option/option_passport.c b/src/game/option/option_passport.c
index 025683a6..e7e703cd 100644
--- a/src/game/option/option_passport.c
+++ b/src/game/option/option_passport.c
@@ -37,11 +37,11 @@ typedef struct PASSPORT_STATUS {
 } PASSPORT_STATUS;
 
 static PASSPORT_STATUS m_PassportStatus = {
-    PAGE_1,
-    PASSPORT_MODE_BROWSE,
-    { false, true, true },
-    { PASSPORT_MODE_UNAVAILABLE, PASSPORT_MODE_NEW_GAME,
-      PASSPORT_MODE_EXIT_TITLE },
+    .page = PAGE_1,
+    .mode = PASSPORT_MODE_BROWSE,
+    .page_available = { false, true, true },
+    .page_role = { PASSPORT_MODE_UNAVAILABLE, PASSPORT_MODE_NEW_GAME,
+                   PASSPORT_MODE_EXIT_TITLE },
 };
 
 static bool m_IsTextInit = false;
@@ -105,6 +105,7 @@ static void Option_PassportInitText(void);
 static void Option_PassportShutdownText(void);
 static void Option_PassportShutdown(INVENTORY_ITEM *inv_item);
 static void Option_PassportDeterminePages(void);
+static int32_t Option_PassportGetNumAvailablePages(void);
 static void Option_PassportInitSaveRequester(int16_t page_num);
 static void Option_PassportInitSelectLevelRequester(void);
 static void Option_PassportInitNewGameRequester(void);
@@ -205,13 +206,17 @@ static void Option_PassportDeterminePages(void)
 
     case INV_LOAD_MODE:
         m_PassportStatus.mode = PASSPORT_MODE_LOAD_GAME;
-        m_PassportStatus.page_available[PAGE_1] = g_SavedGamesCount > 0;
+        m_PassportStatus.page_available[PAGE_1] = true;
         m_PassportStatus.page_available[PAGE_2] = false;
         m_PassportStatus.page_available[PAGE_3] = false;
         m_PassportStatus.page_role[PAGE_1] = PASSPORT_MODE_LOAD_GAME;
         m_PassportStatus.page_role[PAGE_2] = PASSPORT_MODE_UNAVAILABLE;
         m_PassportStatus.page_role[PAGE_3] = PASSPORT_MODE_UNAVAILABLE;
         Option_PassportInitSaveRequester(PAGE_1);
+        if (g_SavedGamesCount == 0) {
+            m_PassportStatus.mode = PASSPORT_MODE_BROWSE;
+            m_PassportStatus.page_role[PAGE_1] = PASSPORT_MODE_NEW_GAME;
+        }
         break;
 
     case INV_SAVE_MODE:
@@ -500,13 +505,26 @@ static void Option_PassportSaveGame(void)
     }
 }
 
+static int32_t Option_PassportGetNumAvailablePages(void)
+{
+    int32_t count = 0;
+    for (int i = PAGE_1; i < PAGE_1 + PAGE_COUNT; i++) {
+        if (m_PassportStatus.page_available[i]) {
+            count++;
+        }
+    }
+    return count;
+}
+
 static void Option_PassportNewGame(void)
 {
     Text_ChangeText(
         m_Text[TEXT_PAGE_NAME], g_GameFlow.strings[GS_PASSPORT_NEW_GAME]);
 
     if (m_PassportStatus.mode == PASSPORT_MODE_BROWSE) {
-        if (g_InputDB.menu_confirm
+        const bool only_ng_available =
+            Option_PassportGetNumAvailablePages() == 1;
+        if ((g_InputDB.menu_confirm || only_ng_available)
             && (g_Config.enable_game_modes
                 || g_Config.profile.new_game_plus_unlock)) {
             Option_PassportInitNewGameRequester();
```

**Comment by rr-** [comment]
I also think that we should change the names of the new cleanup functions - they slightly deviate from the established naming conventions. "Shutdown" usually refers to the release of memory and resources; moreover, we have a public "Init" and currently "Shutdown" only clears internal flags and is private. For better symmetry with "Init" and to better reflect its purpose, I'd change "Shutdown" to "Close" and "FreeRequestors" to "Shutdown".

TLDR a proposed patch:

```diff
diff --git a/src/game/option/option.c b/src/game/option/option.c
index eed3da41..2ed76f7b 100644
--- a/src/game/option/option.c
+++ b/src/game/option/option.c
@@ -18,7 +18,7 @@ void Option_Init(void)
 
 void Option_Shutdown(void)
 {
-    Option_PassportFreeRequestors();
+    Option_PassportShutdown();
 }
 
 void Option_DoInventory(INVENTORY_ITEM *inv_item)
diff --git a/src/game/option/option_passport.c b/src/game/option/option_passport.c
index 025683a6..c9803078 100644
--- a/src/game/option/option_passport.c
+++ b/src/game/option/option_passport.c
@@ -103,7 +103,7 @@ REQUEST_INFO g_SavegameRequester = {
 
 static void Option_PassportInitText(void);
 static void Option_PassportShutdownText(void);
-static void Option_PassportShutdown(INVENTORY_ITEM *inv_item);
+static void Option_PassportClose(INVENTORY_ITEM *inv_item);
 static void Option_PassportDeterminePages(void);
 static void Option_PassportInitSaveRequester(int16_t page_num);
 static void Option_PassportInitSelectLevelRequester(void);
@@ -126,7 +126,7 @@ void Option_PassportInit(void)
         (g_GameFlow.level_count + 1) * m_SelectLevelRequester.item_text_len);
 }
 
-void Option_PassportFreeRequestors(void)
+void Option_PassportShutdown(void)
 {
     Memory_FreePointer(&g_SavegameRequester.item_texts);
     Memory_FreePointer(&m_SelectLevelRequester.item_texts);
@@ -163,7 +163,7 @@ static void Option_PassportShutdownText(void)
     m_IsTextInit = false;
 }
 
-static void Option_PassportShutdown(INVENTORY_ITEM *inv_item)
+static void Optio_PassportClose(INVENTORY_ITEM *inv_item)
 {
     if (m_PassportStatus.page == PAGE_3) {
         inv_item->anim_direction = 1;
@@ -700,11 +700,11 @@ void Option_Passport(INVENTORY_ITEM *inv_item)
             g_Input = (INPUT_STATE) { 0 };
             g_InputDB = (INPUT_STATE) { 0 };
         } else {
-            Option_PassportShutdown(inv_item);
+            Optio_PassportClose(inv_item);
         }
     }
 
     if (g_InputDB.menu_confirm) {
-        Option_PassportShutdown(inv_item);
+        Optio_PassportClose(inv_item);
     }
 }
diff --git a/src/game/option/option_passport.h b/src/game/option/option_passport.h
index d6a28c72..8377b8b2 100644
--- a/src/game/option/option_passport.h
+++ b/src/game/option/option_passport.h
@@ -4,4 +4,4 @@
 
 void Option_Passport(INVENTORY_ITEM *inv_item);
 void Option_PassportInit(void);
-void Option_PassportFreeRequestors(void);
+void Option_PassportShutdown(void);
```

**Comment by walkawayy** [comment]
Agreed about the shutdown conventions.

Regarding hitting F6 with no saves, it should bring up new game in the gym. But, in TombATI, hitting quick load with no saves in a normal level would bring up the save game menu instead, not new game. Tbh, neither of those behaviors make sense to me, but if I had to choose, I think bringing up the save game menu in normals levels like TombATI make more sense to me than new game.

**Comment by rr-** [comment]
Hmm let's hear more opinion on this as it'll be later unwieldy to change yet again. @richard-l @lahm86 wdyt ?

**Comment by Richard-L** [comment]
How about showing the quick load menu as per usual, but just it saying "no saves available yet", or in a similar wording.

Just thinking that that allows the behaviour of what F6 does to stay consistent, which seems desirable to me.

**Comment by lahm86** [comment]
Agree with @Richard-L, that sounds good 👍

**Comment by walkawayy** [comment]
Ok I changed the quick load to show an empty passport menu when there are no saves. It still skips through the page on the title menu or normal inventory screens. And shows new game in the gym. Also reverted the `Option_PassportShutdown` name and renamed the new closing function to `Option_PassportClose`. Lmk if everything looks good.

**Review by rr- — APPROVED** [review]

**Review by lahm86 — APPROVED** [review]
